File: 094 Binary Tree Inorder Traversal/binary_tree_inorder_traversal.cpp
/*	
 *	LeetCode Submissions by Xinyu Chen
 *	Binary Tree Inorder Traversal
 *	https://leetcode.com/problems/binary-tree-inorder-traversal/
 *	Runtime: 0 ms
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class SolutionRecursive {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        inorder(root, result);
        return result;
    }
    
    void inorder(TreeNode* node, vector<int>& result) {
        if (node == nullptr) return;
        inorder(node->left, result);
        result.push_back(node->val);
        inorder(node->right, result);
    }
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode *> st;
        vector<int> res;
        while (root != nullptr || !st.empty()) {
            if (root != nullptr) {
                st.push(root);
                root = root->left;
            } else {
                TreeNode *node = st.top();
                st.pop();
                res.push_back(node->val);
                root = node->right;
            }
        }
        return res;
    }
};  // Author: XC

File: 094 Binary Tree Inorder Traversal/binary_tree_inorder_traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
// Recursive version
public class Solution { // Author: Xinyu Chen
    private List<int> result;
    public IList<int> InorderTraversal(TreeNode root) {
        result = new List<int>();
        inorder(root);
        return result;
    }
    
    private void inorder(TreeNode node) 
    {
        if (node == null) return;
        inorder(node.left);
        result.Add(node.val);
        inorder(node.right);
    }
}
// Iterative version
public class Solution { // Author: Xinyu Chen
    public IList<int> InorderTraversal(TreeNode root) {
        var result = new List<int>();
        var st = new Stack<TreeNode>();
        while (st.Count > 0 || root != null)
        {
            if (root != null)
            {
                st.Push(root);
                root = root.left;
            }
            else
            {
                root = st.Pop();
                result.Add(root.val);
                root = root.right;
            }
        }
        return result;
    }
}
